File: api/bukgetapi/v3.py
import json
import bleach
from bottle import Bottle, redirect, response, request, abort, HTTPError
import common as c

app = Bottle()


@app.hook('before_request')
def set_json_header():
    # We will need these set for everything in the API ;)
    response.set_header('Content-Type', 'application/json')
    response.set_header('Access-Control-Allow-Origin', '*')


@app.get('/')
@app.get('/geninfo')
@app.get('/geninfo/')
def generation_info():
    '''Generation Information
    Returns the generation information as requested.  User can optionall request
    to look X number of versions back.
    '''
    size = c.sint(request.query.size or None)
    callback = request.query.callback or None
    return c.jsonify(c.list_geninfo(size), callback)


@app.get('/geninfo/<idnum>')
def get_geninfo(idnum):
    '''Specific Generation Information
    Return information on a specific generation.
    '''
    callback = request.query.callback or None
    return c.jsonify(c.get_geninfo(idnum), callback)


@app.get('/plugins')
@app.get('/plugins/')
@app.get('/plugins/<server>')
@app.get('/plugins/<server>/')
def plugin_list(server=None):
    '''Plugin Listing
    Returns the plugin listing.  Can optionally be limited to a specific server
    binary compatability type.
    '''
    fields = (request.query.fields or 'slug,plugin_name,description').split(',')
    start = c.sint(request.query.start or None)
    size = c.sint(request.query.size or None)
    sort = request.query.sort or 'slug'
    callback = request.query.callback or None
    data = c.list_plugins(server, fields, sort, start, size)
    return c.jsonify(data, callback)


@app.get('/plugins/<server>/<slug>')
@app.get('/plugins/<server>/<slug>/')
@app.get('/plugins/<server>/<slug>/<version>')
@app.get('/plugins/<server>/<slug>/<version>/')
def plugin_details(server, slug, version=None):
    '''Plugin Details 
    Returns the document for a specific plugin.  Optionally can return only a
    specific version as part of the data as well.
    '''
    fields = (request.query.fields or '').split(',')
    size = c.sint(request.query.size or None)
    data = c.plugin_details(server, slug, version, fields)
    if not data: abort(404, "Plugin Does Not Exist")
    callback = request.query.callback or None
    if size is not None:
        data['versions'] = data['versions'][:size]
    return c.jsonify(data, callback)


@app.get('/plugins/<server>/<slug>/<version>/download')
def plugin_download(server, slug, version):
    '''Plugin Download Redirector
    Will attempt to redirect to the plugin download for the version specified.
    If no version exists, then it will throw a 404 error.
    '''
    plugin = c.plugin_details(server, slug, version, {})
    if version.lower() == 'latest':
        link = [plugin['versions'][0]['download'],]
    else:
        versions = plugin['versions']
        link = [v['download'] for v in versions if v['version'] == version]
    if len(link) > 0:
        redirect(link[0])
    else:
        raise HTTPError(404, '{"error": "could not find version"}')


@app.get('/authors')
@app.get('/authors/')
def author_list():
    '''Author Listing
    Returns a full listing of the authors in the system and the number of
    plugins that they have in the database.
    '''
    callback = request.query.callback or None
    return c.jsonify(c.list_authors(), callback)


@app.get('/authors/<name>')
@app.get('/authors/<name>/')
@app.get('/authors/<server>/<name>')
@app.get('/authors/<server>/<name>/')
def author_plugins(name, server=None):
    '''Author Plugin Listing
    Returns the plugins associated with a specific author.  Optionally can also
    be restricted to a specific server binary compatability.
    '''
    callback = request.query.callback or None
    fields = (request.query.fields or 'slug,plugin_name,description').split(',')
    start = c.sint(request.query.start or None)
    size = c.sint(request.query.size or None)
    sort = request.query.sort or 'slug'
    data = c.list_author_plugins(server, name, fields, sort, start, size)
    return c.jsonify(data, callback)


@app.get('/categories')
@app.get('/categories/')
def category_list():
    '''Category Listing
    Returns the categories in the database and the number of plugins that each
    category holds.
    '''
    callback = request.query.callback or None
    return c.jsonify(c.list_categories(), callback)


@app.get('/categories/<name>')
@app.get('/categories/<name>/')
@app.get('/categories/<server>/<name>')
@app.get('/categories/<server>/<name>/')
def category_plugins(name, server=None):
    '''Category Plugin listing
    returns the list of plugins that match a specific category.  Optionally a
    specific server binary compatability can be specified.
    '''
    callback = request.query.callback or None
    fields = (request.query.fields or 'slug,plugin_name,description').split(',')
    start = c.sint(request.query.start or None)
    size = c.sint(request.query.size or None)
    sort = request.query.sort or 'slug'
    data = c.list_category_plugins(server, name, fields, sort, start, size)
    return c.jsonify(data, callback)


@app.post('/updates')
def get_updates():
    '''Plugin updates
    Returns a list of dictionaries with the latest, release, beta, and alpha
    version numbers.
    '''
    callback = request.forms.get('callback') or None
    slugs = (request.forms.get('slugs') or '').split(',')
    server = request.forms.get('server') or 'bukkit'
    try:
        data = c.plugins_up_to_date(slugs, server)
    except:
        raise HTTPError(400, '{"error": "invalid post"}')
    else:
        return c.jsonify(data, callback)


@app.post('/search')
@app.get('/search/<field>/<action>/<value>')
@app.get('/search/<field>/<action>/<value>/')
def search(field=None, action=None, value=None):
    '''Plugin search
    A generalized search system that accepts both single-criteria get requests
    as well as multi-criteria posts.
    '''
    filters = []
    if request.method == 'GET':
        callback = request.query.callback or None
        fields = (request.query.fields or 'slug,plugin_name,description').split(',')
        start = c.sint(request.query.start or None)
        size = c.sint(request.query.size or None)
        sort = request.query.sort or 'slug'
        field = field
        value = value
        filters = [
            {'field': field, 'action': action, 'value': value}
        ]
    else:
        callback = request.forms.get('callback') or None
        filters = json.loads(request.forms.get('filters') or '[]'
        fields = (request.forms.get('fields')) or 'slug,plugin_name,description').split(',')
        start = c.sint(request.forms.get('start') or None
        size = c.sint(request.forms.get('size') or None
        sort = request.forms.get('sort') or 'slug'
    try:
        data = c.plugin_search(filters, fields, sort, start, size)
    except:
        raise HTTPError(400, '{"error": "invalid post"}')
    else:
        return c.jsonify(data, callback)

File: v3.js
module.exports = function (app, db, common) {

  // This could become a middleware in an ideal reality.
  // 
  // Ecmascript 6 would make this easier...
  // let [ fields, start, size, sort ] = handle_parameters(req);
  function handle_parameters (req, full, callback) {
    if (req.params.server == null) {
      req.params.server = undefined;
    }

    if (req.params.version == null) {
      req.params.version = undefined;
    }

    var fields;
    if (full) {
        fields = ((req.query.fields == null ? '' : req.query.fields.split(',')));
    } else {
        fields = ((req.query.fields == null ? 'slug,plugin_name,description' : req.query.fields).split(','));
    }
    var start = req.query.start == null ? undefined : parseInt(req.query.start);
    var size = req.query.size == null ? undefined : parseInt(req.query.size);
    var sort = req.query.sort == null ? 'slug' : req.query.sort;

    return callback(fields, start, size, sort);
  }

  function geninfo (req, res, next) {
    handle_parameters(req, false, function (fields, start, size, sort) {
      common.list_geninfo(size, function (callback) {
        res.send(callback);
        next();
      });
    });
  }

  function plugin_list (req, res, next) {
    handle_parameters(req, false, function (fields, start, size, sort) {
      common.list_plugins(req.params.server, fields, sort, start, size, function (callback) {
        if (callback == null) {
          res.send(400, {
            'error': 'invalid params'
          });
          return next();
        }
        res.send(callback);
        next();
      });
    });
  }

  function plugin_details (req, res, next) {
    handle_parameters(req, true, function (fields, start, size, sort) {
      common.plugin_details(req.params.server, req.params.slug, req.params.version, fields, function (data) {
        if (data == null) {
          res.send(404, { 'error' : 'Plugin Does Not Exist' });
          return next();
        }

        if (size != undefined && data['versions'] != null) {
          data['versions'].length = size;
        }

        res.send(data);
        next();
      });
    });
  }

  function author_plugins (req, res, next) {
    handle_parameters(req, false, function (fields, start, size, sort) {
      common.list_author_plugins(req.params.server, req.params.name, fields, sort, start, size, function (callback) {
        if (callback == null) {
          res.send(400, {
            'error': 'invalid author'
          });
          return next();
        }
        res.send(callback);
        next();
      });
    });
  }

  function category_plugins (req, res, next) {
    handle_parameters(req, false, function (fields, start, size, sort) {
      common.list_category_plugins(req.params.server, req.params.name, fields, sort, start, size, function (callback) {
        if (callback == null) {
          res.send(400, {
            'error': 'invalid category'
          });
          return next();
        }
        res.send(callback);
        next();
      });
    });
  }

  app.get('/3/', function (req, res, next) {
    geninfo(req, res, next);
  });

  app.get('/3/geninfo', function (req, res, next) {
    geninfo(req, res, next);
  });

  app.get('/3/geninfo/:idnum', function (req, res, next) {
    common.get_geninfo(req.params.idnum, function (callback) {
      res.send(callback);
      next();
    });
  });

  app.get('/3/plugins', function (req, res, next) {
    plugin_list(req, res, next);
  });

  app.get('/3/plugins/:server', function (req, res, next) {
    plugin_list(req, res, next);
  });

  app.get('/3/plugins/:server/:slug', function (req, res, next) {
    plugin_details(req, res, next);
  });

  app.get(/^\/3\/plugins\/(.*)\/(.*)\/(.*)\/download/, function (req, res, next) {
    common.plugin_details(req.params[0], req.params[1], req.params[2], [], function (data) {
      if (data == null) {
        res.send(404, { 'error' : 'Plugin Does Not Exist' });
        return next();
      }

      if (req.params[2].toLowerCase() == 'latest') {
        res.header('Location', data['versions'][0]['download']);
        res.send(302);
        return next();
      } else {
        for (var i = 0, il = data['versions'].length; i < il; i++) {
          if (data['versions'][i]['version'] == req.params[2]) {
            res.header('Location', data['versions'][i]['download']);
            res.send(302);
            return next();
          }
        }
      }

      res.send(404, {'error': 'could not find version'});
      next();
    });
  });

  app.get(/^\/3\/plugins\/(.*)\/(.*)\/(.*)/, function (req, res, next) {
    req.params.server = req.params[0];
    req.params.slug = req.params[1];
    req.params.version = req.params[2];
    plugin_details(req, res, next);
  });


  app.get('/3/authors', function (req, res, next) {
    common.list_authors(function (callback) {
      res.send(callback);
      next();
    });
  });

  app.get('/3/authors/:name', function (req, res, next) {
    author_plugins(req, res, next);
  });

  app.get('/3/authors/:server/:name', function (req, res, next) {
    author_plugins(req, res, next);
  });

  app.get('/3/categories', function (req, res, next) {
    common.list_categories(function (callback) {
      res.send(callback);
      next();
    });
  });

  app.get('/3/categories/:name', function (req, res, next) {
    category_plugins(req, res, next);
  });

  app.get('/3/categories/:server/:name', function (req, res, next) {
    category_plugins(req, res, next);
  });

  app.post('/3/updates', function (req, res, next) {
    var slugs = (req.params.slugs == null ? '' : req.params.slugs).split(',');
    var server = req.params.server == null ? 'bukkit' : req.params.server;
    var hashes = (req.params.hashes == null ? '' : req.params.hashes).split(',');
    var filenames = (req.params.filenames == null ? '' : req.params.filenames).split(',');
    var extra_fields = ((req.params.extra_fields == null ? '' : req.params.extra_fields.split(',')));
    var extra_version_fields = ((req.query.extra_version_fields == null ? '' : req.query.extra_version_fields.split(',')));
    common.plugins_up_to_date(slugs, hashes, filenames, server, extra_fields, extra_version_fields, function (callback) {
      res.send(callback);
      next();
    });
  });

  app.get('/3/updates', function (req, res, next) {
    var slugs = (req.query.slugs == null ? '' : req.query.slugs).split(',');
    var server = req.query.server == null ? 'bukkit' : req.query.server;
    var hashes = (req.query.hashes == null ? '' : req.query.hashes).split(',');
    var filenames = (req.query.filenames == null ? '' : req.query.filenames).split(',');
    var extra_fields = ((req.query.extra_fields == null ? '' : req.query.extra_fields.split(',')));
    var extra_version_fields = ((req.query.extra_version_fields == null ? '' : req.query.extra_version_fields.split(',')));
    common.plugins_up_to_date(slugs, hashes, filenames, server, extra_fields, extra_version_fields, function (callback) {
      res.send(callback);
      next();
    });
  });

  function search (req, res, next) {
    var filters = []

    if (req.method == 'GET') {
      var fields = ((req.query.fields == null ? 'slug,plugin_name,description' : req.query.fields).split(','));
      var start = req.query.start == null ? undefined : parseInt(req.query.start);
      var size = req.query.size == null ? undefined : parseInt(req.query.size);
      var sort = req.query.sort == null ? 'slug' : req.query.sort;
      var field = req.params[0];
      var value = req.params[2];
      filters = [{
        'field': field,
        'action': req.params[1],
        'value': value
      }];
    } else {
      var filters = req.params.filters == null ? [] : JSON.parse(req.params.filters);
      var fields = ((req.params.fields == null ? 'slug,plugin_name,description' : req.params.fields).split(','));
      var start = req.params.start == null ? undefined : parseInt(req.params.start);
      var size = req.params.size == null ? undefined : parseInt(req.params.size);
      var sort = req.params.sort == null ? 'slug' : req.params.sort;
    }

    common.plugin_search(filters, fields, sort, start, size, false, function (callback) {
      if (callback == null) {
        res.send(400, {
          'error': 'invalid search'
        });
        return next();
      }

      res.send(callback);
      next();
    });
  }

  app.post('/3/search', function (req, res, next) {
    search(req, res, next);
  });

  app.get(/^\/3\/search\/([a-zA-Z0-9_\.~-]+)\/(\=|\!\=|\<|\<\=|\>|\>\=|[a-z]+)\/([a-zA-Z0-9_\.~-]+)/, function (req, res, next) {
    search(req, res, next);
  });
}